[DO NOT MERGE] Fix Edm.Guid mutator

lubosmj opened the pull request:
This commit should prevent undesirable replacements of immutable characters within a Guid string. Before this commit, the following code snippet raises an error:

```python
from odfuzz.mutators import GuidMutator
while True:
    guid = GuidMutator.replace_char("guid'dddddddd-dddd-dddd-dddd-dddddddddddd'")
    if not guid.startswith("guid'") or not guid.endswith("'"):
        raise RuntimeError("oopsie")
```

lubosmj commented:
This PR should prevent replacing of undesirable characters in a string. For example, the following code snippet raises an error almost immediately.

```python
from odfuzz.mutators import GuidMutator
while True:
    guid = GuidMutator.replace_char("guid'dddddddd-dddd-dddd-dddd-dddddddddddd'")
    if not guid.startswith("guid'") or not guid.endswith("'"):
          raise RuntimeError("oopsie")
```

filak-sap commented:
Hi Lubos, does odfuzz have a test suite?
What about to use your latest PR comment to as the commit message?

lubosmj commented:
@filak-sap, I deleted the comment and moved its content to the commit message.

Anyway, can you elaborate on your first question? If I understand it correctly, you want me to write a new test case for the Edm.Guid mutator proving that the fix really works, is that right? However, I would probably end up writing a test case like this: https://github.com/SAP/odfuzz/blob/master/tests/test_mutator.py#L31-L72. I reckon it is not a good way how to test the mutators. We should rather open a new issue and discuss different possibilities of creating more robust tests which can handle, for example, the non-deterministic behavior.

phanak-sap commented:
@lubosmj - well, this is fine example of what I tried to explain to you personally, that the project is in current state not easy to test. We should take this opportunity to design more testable API from bottom-up. 

There are few ways how to test code as in `GuidMutator.replace_char`, but since what it does is string replacement and the random number is used in one place, this should be simply rewritten to be deterministic and testable method (instead of usage of something like testing in loops (like your first comment) or for example property testing, e.g. Hypothesis library).

Especially if this is fix for issue #40  - first you should try to write test that reproduces the bug, if it is possible. 

As a quick, simple -but "hey, it is working" solution can be adding optional parameter to the method and if present, it will set the value for `rand_hex_char` instead of random number. And then you shoudl be able to write parametrized pytests for every possibilty, some of which should fail and therefore reproduce this bug.

Please create separate PR for this change. It expected to be failed on travis and I want to see such build log. 

Then add fix from this commit there, proving that it fixes the problem uncovered by the test, and I will close this as duplicate.

And then we will move similarly for other odata data types.

lubosmj commented:
@phanak-sap, the commit is not a fix for #40. I submitted the pull request #45 which should convince you that this is a real problem and it should be resolved as soon as possible in order to prevent unnecessary server errors during fuzzing.

In my opinion, once I add a new function parameter to the mutator, it may decrease the performance due to the fact that ODfuzz is going to check for that passed parameter always. We want to avoid such checks because the fuzzing process needs to be fast. I am in favor of creating a different API for the mutators globally. It really does not make sense to pass `rand_hex_char` to the function and create a wrapper around it because this will have to be done for every single mutator out there then. What about creating a new issue where we can discuss what kind of API should be the best to use?

phanak-sap commented:
OK, I have created blank issue #46 to discuss this topic and tie PRs to.

filak-sap commented:
Guys, the initial comment contains a code snippet which can be used for testing or am I wrong? Why do you spend time writing code snippets instead for writing tests?

phanak-sap commented:
@filak-sap The snipped is endless loop - wait till the bug surfaces. The true, simple and deterministic test for this exact bug is part of already merged #45, with the code fix present in this one. This PR therefore serves no purpose anymore and was closed.
